### PR TITLE
Feature: Support RetryStrategy/InstanceGroups

### PR DESCRIPTION
**Pull request by ananth102**
Note: Some modifications are made to Hyper parameter tuning, this is normal as Hyper parameter tuning consists of training jobs.

Description of changes:
generator.yaml - commented out shapes. Added late initialize to ResourceConfig.InstanceCount and it's HPO equivalent. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: generator.yaml
Comment (surajkota) [inline generator.yaml:191]: is this change in behaviour?

what happens if user is using instance groups only?

**Review by surajkota — COMMENTED** [review]
shall we add the retryStrategy to existing e2e tests? free of cost fault tolerance

File: generator.yaml
Comment (ananth102) [reply generator.yaml:191]: Added retry strategy to integration tests and added unit tests.

>is this change in behavior?
Yes, instance count is optional because the user can specify Instance group. However the server side default of instanceCount is 0.
> what happens if user is using instance groups only?
Instance count will be late initialized to 0.

**Review by surajkota — CHANGES_REQUESTED** [review]
missing handling `TrainingJobDefinitions`.ResourceConfig.InstanceCount in HPO

also reduce the retry count to 3

File: generator.yaml
Comment (surajkota) [reply generator.yaml:191]: this is a very strange behaviour

**Comment by ananth102** [comment]
> missing handling TrainingJobDefinitions.ResourceConfig.InstanceCount in HPO

ACK, I've been trying to test that but I keep getting credential errors on sagemaker.

**Comment by ananth102** [comment]
/hold
Merge this after #177. Also still need to debug TrainingJobDefinitions.ResourceConfig.InstanceCount.

**Comment by ananth102** [comment]
/unhold
TrainingJobDefinitions.ResourceConfig.InstanceCount works

**Comment by surajkota** [comment]
/ok-to-test

looks ok, do you want to merge 177 first?

**Comment by ananth102** [comment]
\hold

**Review by ryansteakley — APPROVED** [review]

**Pull request merged**
